### PR TITLE
#164798255 Enable users to view and edit their profiles

### PR DESCRIPTION
**Pull request by e-ian**
#### What does this PR do?
- Enable authenticated users to view and edit their profiles
#### Description of tasks to be completed.
- Add tests for profile and edit profile functionality
- Add functionality for getting a user's profile
- Add functionality for a user to edit their profile
#### How should this be manually tested?
- Visit the deployed application page
- Login using the registered email and password using the sample credentials below:
`email: ianemma70@gmail.com`
`password: @bochiSupreme1`
- On successful login, a toast message shows that a successful login and then you are redirected to the home/landing page with the `profile` link the navbar.
- Visit the profile page to view the current logged user's profile
- To edit a profile, click on edit profile and edit your `full_name`, `bio` and `image`.
- On a successful edit, a toast appears and you are redirected to the profile page
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[#164798255](https://www.pivotaltracker.com/story/show/164798255)

#### Screenshots(if appropriate)
- profile page
image

- Edit profile page
image

### CODE REVIEW

**Review by anyatibrian — APPROVED** [review]
great work,i really liked your implementations of the profile view

**Comment by e-ian** [comment]
> great work,i really liked your implementations of the profile view

Thank you @anyatibrian

File: src/components/routes.js [hunk 1]
```diff
@@ -4,18 +4,20 @@ import LoginContainer from "../containers/Login/LoginContainer";
...
     return (
       <BrowserRouter>
-        <Switch>
-          <Route path="/" component={Home} exact={true} />
-          <Route path="/register" component={RegisterContainer} />
-          <Route exact path="/home" component={Home} />
-          <Route path="/login" component={LoginContainer} />
-          <Route path="/socialAuth" component={SocialAuthContainer} />
-        </Switch>
+        <Route path="/" component={Home} exact={true} />
```
Comment (patrickf949) [inline src/components/routes.js]: Great work @e-ian 
I see you removed the `<Switch>`
It is actually important so as to get the first matching url as opposed to all urls that may match the same link.
For example  `/profile/edit` would match `/profile` and `/profile/edit` without the `<Switch>`
With `<Switch>`  it only matches `/profile/edit`

Comment (e-ian) [reply src/components/routes.js]: Thanks @patrickf949 . I have worked on this issue. It is now fine.

**Review by patrickf949 — APPROVED** [review]
Great work @e-ian LGTM

**Review by Genza999 — APPROVED** [review]
Nice work @e-ian

**Review by MuhanguziDavid — CHANGES_REQUESTED** [review]
Nice work @e-ian , however, I can still see a profile link on the navbar when I am logged out. I can also see links to signup and login on the navbar when I am already logged in. Fix this to ensure that `profile` only appears when one is logged in and `sign in` and `get started` only appear when one is logged out.

Screen Shot 2019-05-17 at 10 29 39

File: src/components/userProfile/EditProfile.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      history.push("/profile");
+    }
+  }
+
+  componentWillMount() {
+    const { full_name, bio, image } = this.props.profile;
+    this.setState({ full_name, bio, image });
+  }
+  handleNameChange(event) {
+    this.setState({ full_name: event.target.value });
```
Comment (MuhanguziDavid) [inline src/components/userProfile/EditProfile.js]: please, destructure this

File: src/components/userProfile/EditProfile.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  handleBioChange(event) {
+    this.setState({ bio: event.target.value });
```
Comment (MuhanguziDavid) [inline src/components/userProfile/EditProfile.js]: destructure this as well

File: src/components/userProfile/EditProfile.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  handleImageUpload(event) {
+    this.setState({ image: event.target.files });
+  }
+  handleSubmit(event) {
+    event.preventDefault();
+    const data = {
+      bio: this.state.bio,
+      full_name: this.state.full_name,
```
Comment (MuhanguziDavid) [inline src/components/userProfile/EditProfile.js]: destructure this as well. Ensure to go through the file and destructure where were necessary

File: src/containers/profile/ProfileContainer.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      username: "",
+      image: "",
+      bio: "",
+      full_name: "",
+      isUploading: ""
+    };
+  }
+
+  componentWillMount() {
+    this.props.getProfileAction();
```
Comment (MuhanguziDavid) [inline src/containers/profile/ProfileContainer.js]: go through the file and destructure where necessary

File: src/tests/components/EditProfile.test.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    const event = {
+      target: {
+        name: "image",
+        files: "files.jpg"
+      }
+    };
+    wrapper.instance().handleImageUpload(event);
+    expect(wrapper.instance().state.image).toBe("files.jpg");
+  });
+  it("should simulate image upload", () => {
```
Comment (MuhanguziDavid) [inline src/tests/components/EditProfile.test.js:78]: You have to expect or assert something in your test

File: src/components/userProfile/EditProfile.js (hunk 3 above)
Comment (e-ian) [reply src/components/userProfile/EditProfile.js]: I have done this.

File: src/components/userProfile/EditProfile.js (hunk 4 above)
Comment (e-ian) [reply src/components/userProfile/EditProfile.js]: Done

File: src/containers/profile/ProfileContainer.js (hunk 5 above)
Comment (e-ian) [reply src/containers/profile/ProfileContainer.js]: Done

File: src/tests/components/EditProfile.test.js (hunk 6 above)
Comment (e-ian) [reply src/tests/components/EditProfile.test.js:78]: Done

File: src/components/userProfile/EditProfile.js (hunk 2 above)
Comment (e-ian) [reply src/components/userProfile/EditProfile.js]: Done

**Comment by e-ian** [comment]
> Nice work @e-ian , however, I can still see a profile link on the navbar when I am logged out. I can also see links to signup and login on the navbar when I am already logged in. Fix this to ensure that `profile` only appears when one is logged in and `sign in` and `get started` only appear when one is logged out.
> 
> Screen Shot 2019-05-17 at 10 29 39

I have fixed this issue @MuhanguziDavid

**Review by MuhanguziDavid — CHANGES_REQUESTED** [review]
Nice work @e-ian , however, I notice that the navbar still shows `sign in` and `get started` when i login using social auth.

I also noticed that when I click `Authors Haven` in the navbar to take me back to the homepage, `sign up` and `get started` appear in the nav bar. Please fix this.
You can use the `username` session variable as criteria for the switch in the navbar and add a logout button which clears the session and redirects to the landing page.

**Comment by e-ian** [comment]
> Nice work @e-ian , however, I notice that the navbar still shows `sign in` and `get started` when i login using social auth.
> 
> I also noticed that when I click `Authors Haven` in the navbar to take me back to the homepage, `sign up` and `get started` appear in the nav bar. Please fix this.
> You can use the `username` session variable as criteria for the switch in the navbar and add a logout button which clears the session and redirects to the landing page.

I have implemented this feedback @MuhanguziDavid

**Pull request merged**